doc(ssdlc): Add hardening page (#6297)
canonical/cloud-init: doc/rtd/explanation/hardening.rst

Security Hardening
******************

Cloud-init's use case is automating cloud instance initialization, with support
across distributions and platforms. There is a myriad of ways to harden this
space.

Follow the security hardening guidelines provided by the OSes and cloud

> platforms that your cloud-init configuration is targeting.

Reviewer: It may be worth stating that many cloud platforms provide SSH public keys in metadata which setup the default user with the appropriate configured means of access using SSH public/private key pairs. 

Most of the harmful security exposure comes when custom user-data presented as #cloud-config or run scripts by the end-user at VM launch time which provides credentials in the form of clear passwords or credentials encoded in URLs for services.

We may  want to recommend that people don't provide clear-text passwords or credentials in any runcmd/bootcmd or user-data scripts `#/bin/bash` etc because that config user-data may be accessible to others on a local network based on a given cloud platform's instance metadata service and instead obtain those credentials to service-endpoints from a vault service or configuration management service such as puppet/chef/ansible/saltstack etc.

> No plain text passwords

Reviewer: Is it worth also calling out that it makes sense to ensure we have updated packages in your VM images upon launch to ensure latest fixes available are actually installed within a potentially stale cloud image? While Ubuntu cloud images are configured by default to enable unattended-upgrades (and resolve this issue overnight) it does leave a gap in current days launches that some software will remain out of date until packages are upgraded.

```
#cloud-config
package_upgrade: true 
```

We could link to Ubuntu's security guidelines https://documentation.ubuntu.com/server/explanation/security/security_suggestions/#keep-your-system-up-to-date

Reviewer 2: > it makes sense to ensure we have updated packages in your VM images upon launch to ensure latest fixes available are actually installed within a potentially stale cloud image?

Recommending an update is fine, but I would suggest that we recommend using current images rather than assuming/accepting outdated ones. Launching an image with the latest packages is more secure than upgrading from an outdated image.

Author: I think reality is a bit in the middle, as (IIRC) there are clouds where the image publishing cadence is longer than months (like Oracle or IBM clouds).

=======================

> While creating users with the

Reviewer: Let's provide a preferred alternative to plain_text_passwds. (and passwords in general).

Do we want to tell folks to instead use `hashed_passwd` and appropriate rounds of salt?

Shall we instead suggest that folks use SSH public keys  via `ssh_import_id` or  `ssh_authorized_keys` instead and point ot security-related docs in Ubuntu for that suggestion?
https://documentation.ubuntu.com/server/how-to/security/openssh-server/#ssh-keys

:ref:`Users and Groups module <mod_cc_users_groups>`, do not use the

> `user.plain_text_passwd` key with its associated value as plain text.

Reviewer: Double quotes to bold instead of italics
```suggestion
``user.plain_text_passwd`` key with its associated value as plain text.
```

> As anyone with access to the configuration might have access to the password.

Reviewer: Another suggestion for hardening security of launched VMs in cloud-init is to document that cloud-init publishes the SSH host public key generated to the serial console which can be validated prior to any SSH client connection to the launched VM. It's give a sense of security that you are actually attaching to the VM you thought you launched and not a MITM.

Security Hardening
******************

Cloud-init's use case is automating cloud instance initialization, with support

> across distributions and platforms. There is a myriad of ways to harden this
> space.

Reviewer: That second sentence is vague. "harden this space" doesn't hold much meaning. Let's correct this to:
```suggestion
across distributions and platforms. There are a myriad of ways to improve the
security posture of a cloud-init configured machine.
```

Follow the security hardening guidelines provided by the OSes and cloud
platforms that your cloud-init configuration is targeting.

Many cloud platforms provide SSH public keys in metadata which setup the
default user with the appropriate configured means of access using
SSH public/private key pairs.

Updated packages
================

To ensure the available security fixes are applied to you VMs images upon

> launch, it is recommended to update the packages

Reviewer: Let's make this a link in RST instead of spelling out the URL inline below 
```suggestion
launch, it is recommended by `Ubuntu security team guidelines` to update the packages
```

.. note::

  Ubuntu cloud images are configured by default to enable unattended-upgrades,
  thus this is resolved this issue when the update gets triggered. One can

>   still apply this recommendation to cloud that gap and update the packages

Reviewer: ```suggestion
  still apply this recommendation to close that gap and update the packages
```

  on first boot.

.. code-block:: yaml

  #cloud-config
  package_update: true
  package_upgrade: true

> Ubuntu's security guidelines: https://documentation.ubuntu.com/server/explanation/security/security_suggestions/#keep-your-system-up-to-date

Reviewer: Let's drop this as an embeded link inline above the section and provide the URL in the footer in RST
```suggestion
```

No plain text passwords
=======================

Most of the harmful security exposure comes when custom user-data presented
as ``#cloud-config`` or run scripts by the end-user at VM launch time which
provides credentials in the form of clear passwords or credentials encoded in
URLs for services.

> It is advised to not provide clear-text passwords or credentials in any
> ``runcmd`` / ``bootcmd`` or user-data scripts #/bin/bash, etc because that
> config user-data may be accessible to others on a local network based on a
> given cloud platform's instance metadata service and instead obtain those
> credentials to service-endpoints from a vault service or configuration
> management service such as Puppet, Chef, Ansible, Saltstack, etc.

Reviewer: ```suggestion
It is recommended not to include plain-text passwords or credentials in any
``runcmd``, ``bootcmd``, or user-data scripts (e.g., ``#!/bin/bash``), as this
configuration user-data may be accessible to others on the local network
depending on the cloud platform's instance metadata service.
Instead, retrieve credentials for service endpoints from a secure
vault or configuration management service such as Puppet, Chef, Ansible,
or SaltStack.
```

While creating users with the
:ref:`Users and Groups module <mod_cc_users_groups>`, do not use the
``user.plain_text_passwd`` key with its associated value as plain text.
``hashed_passwd`` is a more secure alternative.

Avoid plain text passwords with the
:ref:`Set Passwords <mod_cc_set_passwords>`.

Alternatives to user passwords
------------------------------

We recommend using the :ref:`SSH module <mod_cc_ssh>` with ``ssh_import_id`` or
``ssh_authorized_keys`` to import public SSH keys.

> More info:
> https://documentation.ubuntu.com/server/how-to/security/openssh-server/#ssh-keys

Reviewer: ```suggestion
More info from Canonical on `managing SSH-keys for openssh-server`_
```
